.buildkite/pipeline_generator/pipeline_generator.py

```diff
@@ -0,0 +1,180 @@
+    if step.num_nodes > 1:
+        current_step.commands = get_multi_node_test_command(
+            step.commands, 
+            step.working_dir, 
+            step.num_nodes, 
+            step.num_gpus, 
+            get_image_path()
+        )
+        current_step.plugins = None
+
+    def step_should_run():
+        """Determine whether the step should automatically run or not."""
+        if not step.source_file_dependencies or run_all == "1" or step.gpu == "a100":
+            return True
+        return any(source_file in diff_file 
+            for source_file in step.source_file_dependencies 
+            for diff_file in list_file_diff)
+
+    # If step does not have to run automatically, create a block step
+    # and have the current step depend on it
+    if not step_should_run():
+        block_step = get_block_step(step.label)
+        steps.append(block_step)
+        current_step.depends_on = block_step.key
+    steps.append(current_step)
+    return steps
+
+def generate_build_step() -> BuildkiteStep:
+    """Build the Docker image and push it to ECR."""
+    buildkite_commit = os.getenv("BUILDKITE_COMMIT")
+    docker_image = get_image_path()
+    build_commands = [
+        "aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin public.ecr.aws/q9t5s3a7",
+        f"docker build --build-arg max_jobs=16 --build-arg buildkite_commit={buildkite_commit} --build-arg USE_SCCACHE=1 --tag {docker_image} --target test --progress plain .", 
+        f"docker push {docker_image}"
+    ]
+    step = BuildkiteStep(
+        label=":docker: build image", 
+        key="build", 
+        agents={"queue": AgentQueue.AWS_CPU.value}, 
+        env={"DOCKER_BUILDKIT": "1"}, 
+        retry={"automatic": [{"exit_status": -1, "limit": 2}, {"exit_status": -10, "limit": 2}]}, 
+        commands=build_commands,
+        depends_on=None,
+    )
+    return step
+
+def get_external_hardware_tests(test_steps: List[TestStep]) -> List[Union[BuildkiteStep, BuildkiteBlockStep]]:
+    """Process the external hardware tests from the yaml file and convert to Buildkite steps."""
+    with open(EXTERNAL_HARDWARE_TEST_PATH, "r") as f:
+        content = yaml.safe_load(f)
+    buildkite_steps = []
+    for step in content["steps"]:
+        # Fill in AMD Docker image path
+        step["commands"] = [cmd.replace("DOCKER_IMAGE_AMD", get_image_path(AMD_REPO)) for cmd in step["commands"]]
+
+        # Convert step to BuildkiteStep object
+        buildkite_step = BuildkiteStep(**step)
+        buildkite_step.depends_on = "bootstrap"
+
+        # Add block step if step in block list
+        # and have the current step depend on it
+        if buildkite_step.key in STEPS_TO_BLOCK:
+            block_step = get_block_step(buildkite_step.label)
+            buildkite_steps.append(block_step)
+            buildkite_step.depends_on = block_step.key
```

Comment:
this part is repeated?

and do we really want each skipped step/job to have their own "block" step?